Add contributor guidelines & code of conduct update (#679)
typelevel/typelevel.github.com: src/projects/contributing-guide.md

# Contributing to Typelevel Projects

These guidelines apply when contributing to any Typelevel organization project. Please also refer to individual project guidelines for additional guidelines. 

> Typelevel affiliate projects are also welcome to integrate these guidelines in their contributing guidelines.

Reviewer: ```suggestion
Typelevel affiliate projects are also welcome to integrate these guidelines into their contributing guidelines.
```

> This document focuses on policy. See also the [Start Contributing](../community/start-contributing.md) guide if you want step-by-step guidance on contributing to Typelevel projects.

Welcome to the Contributing to Typelevel Projects guide. This document provides an overview of Typelevel's expectations when contributing to organization projects. At a very high level:

We **want** contributors who are interested in gaining a deeper understanding of the projects they're working on, learning how and why they were built, and working with the community to improve our projects.

We do **not** want contributors that increase the workload for maintainers without any gain for the project or any learning for the contributor.

The following sections provide guidance on how to make contributions that fall into the first category.

## Respect our Code of Conduct

When you contribute to Typelevel, we expect that you respect our [Code of Conduct](/code-of-conduct/README.md).

## Use Your Own Words

> Write prose, including issue and pull request descriptions, documentation and code comments in your own words. Do not use automated tools to generate these for you. These artifacts are created to be read by humans, so we believe they should also be written by humans.

Reviewer 2: I'd like to point out again, that source code is _also_ created to be read by humans. ("Programs must be written for people to read, and only incidentally for machines to execute" from SICP.)

While Typelevel communications are primarily conducted in English, our contributors and users are worldwide and may not be proficient in written English. Using machine translation tools (including LLM translation tools) to translate your writing to English is always acceptable. You may also include the original text in your message.

## Start With an Issue

> To contribute to a Typelevel project, start a conversation before creating a pull request. This can be in a new issue, continuing the conversation in an existing issue, or any other communication channel being used by the project. This gives maintainers and other contributors a chance to work with you on your idea at an earlier stage to make sure it is an acceptable contribution.

Reviewer 3: does plugging the typelevel discord make sense here? or link to a resource listing communication channels

Author: I don't really want to have another discord link that might have to be hunted down to update if the link needs to be changed. Also, the right communication channel is usually the project's issues page, which is separate for every project. In the diataxis taxonomy, this in intended to be a reference for the appropriate way to contribute, not a how-to guide (which the start contributing page more functions as).

Reviewer 4: Hmm.  Not sure I 100% agree -- I mean, this paragraph *is* providing how-to information.  And given that most projects have individual channels in the Discord (and some specifically have dev channels), I suspect that it would sometimes be appropriate to start with an "is this even a good idea?" conversation in the Discord.

So a "linked at the bottom of this page" reference might be appropriate here.

Reviewer: > discord link that might have to be hunted down to update if the link needs to be changed

This is something we should fix/address with Laika. We can set the Discord invite url once as a global config variable and reference it in all the places.

> When you are reporting a bug, your issue should clearly state the expected versus actual behavior you are observing. An ideal bug report includes a minimal reproducer that can easily be run (e.g. a scala-cli script, or small repository) to demonstrate the issue. When you are unable to do so, please provide what other information you can, for example: specific steps that caused the issue, stack traces, or error messages.

Reviewer 4: Not certain about this, but seems like a valid and convenient option sometimes:

```suggestion
When you are reporting a bug, your issue should clearly state the expected versus actual behavior you are observing. An ideal bug report includes a minimal reproducer that can easily be run (e.g. a scala-cli script, a small repository, or a saved Scastie) to demonstrate the issue. When you are unable to do so, please provide what other information you can, for example: specific steps that caused the issue, stack traces, or error messages.
```

> When you are reporting a security issue, *do not* open a Github issue. See the project security policy or the [Typelevel security policy](https://typelevel.org/security.html).

Reviewer: Hmm, does this not work?

```suggestion
When you are reporting a security issue, *do not* open a Github issue. See the project security policy or the [Typelevel security policy](/security.md).
```

Author: I have no idea why I did it like that.

## Pull Requests

The following guidelines are particularly relevant for pull requests:

### Respect Copyright

You must be mindful of the copyright and patent rights of anything you submit.

> If you authored every part of your contribution and own the rights, this is not a problem. You can submit your contribution without reading further.

Reviewer: Since we are stating all of this, we should probably mention the case where a contributor's employer owns the rights and that they should do whatever Corporate Due Diligence™️ is needed to ensure they can legally contribute it.

Author: I don't love any of the language I've tried to add this. Practically, I'd expect professional software developers to be aware of those issues.

> However, any code or assets you took from somewhere else, including code generated by AI, may be subject to copyright or patent rights, which you must respect. In such cases, you need to check the license of the material.

Reviewer 4: > including code generated by AI

I would bet that we're going to get arguments about this, given that current Copyright Office policy says otherwise, and that there's basically no practical way to *know* about copyright or patent rights over AI contributions.

This is a deeply messy topic, nowhere near settled in terms of legislation or litigation, and I have mixed feelings about whether it's worth saying this outright, or just leave the point implicit.

Author: I left this part verbatim from the Godot guidelines. I have no strong feelings about keeping or removing it. Though I'd be surprised if AI generated code get a free pass on patents (are software patents still a thing in the US? I forget.)

Reviewer 4: > are software patents still a thing in the US? I forget.

They are, sadly.  I'm not aware of any settled case law around the intersection of software patents and AI yet, but IANAL, so I dunno how that factors in.

Reviewer 4: Decided to have Kagi Research (my general-research tool of choice) do some quick summarization, which resulted in a fairly interesting thread: https://assistant.kagi.com/share/d21a11b0-19a9-4abd-9f11-e59ac3b18da9

tl;dr -- software patents are still a thing, yes, and "generated by AI" is probably *irrelevant* when it come to violation: patented is patented.  AIs probably can't *create* patents, but when they violate them, it's probably exactly the same as when a human does so.

(Note: I hate software patents with a burning passion and wish that whole notion in law could be just plain expunged.  I'm still vaguely embarrassed that my father's boss apparently invented the idea, many decades ago, and moreso that I have my name on several.  But yes, it's something we have to live with.)

How does that affect contributions?  IMO it's a minefield.  Violations tend to be accidental (unlike copyright), and the average developer has no way to know if they're violated patents.  It's a whole field of law that is heavily governed by security-through-obscurity: you're only in trouble if somebody *notices* that you've violated a patent, and AIUI you're generally better off not trying very hard to figure out if you did so.  (Since if you do the research and find that you *have* arguably trod on an existing patent, now you have evidence subject to discovery of intentional violation.)

So while I care a lot that contributors not *intentionally* violate patent, it isn't even remotely obvious how far we should go beyond that, beyond the statement of "don't do that".  The whole framework of software patents is pretty hostile to open source.  (And indeed, to software development generally.)

And all of this is still somewhat hypothetical: there isn't a lot of settled case law around it yet.

If any members of the @typelevel/board have thoughts here, they might be useful.

> "Source-available" is **not** "open-source", and not all open-source licenses are compatible with each other. We strongly recommend against reading any "source-available" or incompatibly licensed code before contributing to Typelevel projects.

Reviewer: ```suggestion
"Source-available" is **not** "open-source", and not all open-source licenses are compatible with each other. We strongly recommend against reading any "source-available" or incompatibly-licensed code before contributing to Typelevel projects.
```

### Contribute One Change at a Time

Each pull request should contain a single self-contained change. Avoid pushing multiple unrelated changes in the same pull request.

> As a rule of thumb, if your pull request could be split into two pull requests without breaking anything, it should probably be two separate pull requests.

Reviewer: Could also mention that stacked PRs are useful if you have multiple changes that are dependent on each other.

https://github.blog/changelog/2026-07-30-stacked-pull-requests-are-now-in-public-preview/

Reviewer: Oh, just saw your comment about _what_ to do without getting into the specifics/mechanics, which I agree with. So maybe nevermind.

The exception to this rule is "batching changes", where you make the same kind of change to multiple places in the codebase. In this case, it is preferable to make one larger pull request instead of many small ones.

> **Note:** Imagine yourself in the shoes of the reviewer: how can you make the PR easy to review and approve?
>      
> This is best achieved if your pull request is simple, coherent, and uncontroversial.
>

> > Limiting pull requests to one change at a time also simplifies the Git history, makes it possible to revert or cherry-pick specific changes, and reduces the potential of accidentally introducing bugs.

Reviewer: Also bisect!

Your pull request description should, as clearly and succinctly as possible, in your own words, explain all the necessary information to understand the changes you made.

For example, if you're just fixing a typo, a single sentence description is appropriate. However, if your change is large, we expect you to be more thorough with your explanation.

Here are the components of a thorough pull request description:

- **Summary of changes:** A short overview of what is changed.
- **Motivation:** Why you opened the pull request. Ideally, this is a link to an issue.
- **Related work:** Link to similar pull requests or ongoing discussions that provide additional context.
- **Technical overview:** Briefly explain each of the changes in this pull request, and why they are necessary.
- **Testing:** How you tested the pull request, and the results of your testing.
- **Discussion:** How you see the pull request in context. For example, list risks and caveats and how they could be mitigated. If existing projects may be negatively affected, you should disclose this here, with particular attention to possible compatibility breakages and regressions.

> - **Additional work:** If applicable, highlight anything you need help with or feedback on, and describe any necessary or potential follow-up work.

Reviewer 3: I'd encourage the specific usage of Github comments to help localize and contextualize the thing that needs extra review power/feedback/teaching

Author: I'd rather not go too deep into the mechanics of how Github PRs work. This document is talking about *what* you should do, not how exactly to do it.

> **Note:** You may not need *every* one of these components, and you do not need to copy this exact structure. Always ask yourself what would be most helpful for reviewers, and try to balance brevity and thoroughness appropriately.

### Contribute Only What You Stand Behind

Only submit code that you understand, are prepared to explain to a maintainer, and be responsible for.

This especially applies if you implement the idea of another person, copy code from elsewhere, or if you use AI to assist you with your contribution. In all of these cases, you must disclose which part of your submission wasn't fully authored by you. Regardless of the code's origin, by opening a pull request, you are putting your reputation as a contributor behind the code.

---

>  This document is based on the [Godot Pull request rules and guidelines](https://github.com/godotengine/godot-contributing-docs/blob/f46474c855c36883b61d0438414365c771ea78dd/pull_requests/pull_request_guidelines.rst) by the Godot Engine Contributors and is licensed under the [CC BY 3.0 license](https://creativecommons.org/licenses/by/3.0/deed.en).

Reviewer: ```suggestion
This document is based on the [Godot Pull request rules and guidelines](https://github.com/godotengine/godot-contributing-docs/blob/f46474c855c36883b61d0438414365c771ea78dd/pull_requests/pull_request_guidelines.rst) by the Godot Engine Contributors and is licensed under the [CC BY 3.0 license](https://creativecommons.org/licenses/by/3.0/deed.en).
```

Only submit code that you understand, are prepared to explain to a maintainer, and be responsible for. This especially applies if you implement the idea of another person, copy code from elsewhere, or if you use AI tools to assist you with your contribution. In all of these cases, you must disclose which part of your submission wasn't fully authored by you. Regardless of the code's origin, by opening a pull request, you are putting your reputation as a contributor behind the code.

#### Use of AI Tools

You may use AI tools to help write code, but must write issue and pull request descriptions, documentation, code comments and other explanatory text in your own words. You are expected to understand the code you contribute well enough to write about the code yourself, regardless of how it was created.

#### Use of Translation Tools

While Typelevel communications are primarily conducted in English, our contributors and users are worldwide and may not be proficient in written English. Using machine translation tools (including LLM translation tools) to translate your writing to English is always acceptable. You may also include the original text in your message.

---

> This document is based on the [Godot Pull request rules and guidelines](https://github.com/godotengine/godot-contributing-docs/blob/f46474c855c36883b61d0438414365c771ea78dd/pull_requests/pull_request_guidelines.rst) by the Godot Engine Contributors and is licensed under the [CC BY 3.0 license](https://creativecommons.org/licenses/by/3.0/deed.en).

Reviewer: Looking at the terms ... https://creativecommons.org/licenses/by/3.0/deed.en#terms

1. We may want to more explicitly call out that we made modifications.
2. Actually, we just need to state that the original work is CC BY 3.0. Our adaptation can be under CC BY 4.0 which is already the website's license, since there is no share-alike requirement.

Author: 1. That was the intent of "based on". The CC guide suggestions for such [are pretty light](https://wiki.creativecommons.org/wiki/Recommended_practices_for_attribution#This_is_a_great_attribution_for_when_you_have_created_an_adaptation)

  > This work, "90fied", is adapted from "[Creative Commons 10th Birthday Celebration San Francisco](http://www.flickr.com/photos/sixteenmilesofstring/8256206923/in/set-72157632200936657)" by [Timothy Vollmer](http://www.flickr.com/photos/sixteenmilesofstring/), used under [CC BY 4.0](http://creativecommons.org/licenses/by/4.0/). "90fied" is licensed under [CC BY 4.0](http://creativecommons.org/licenses/by/4.0/) by [Your name here].

2. Ah so like this?
```suggestion
This document is based on the [Godot Pull request rules and guidelines](https://github.com/godotengine/godot-contributing-docs/blob/f46474c855c36883b61d0438414365c771ea78dd/pull_requests/pull_request_guidelines.rst) by the Godot Engine Contributors which is licensed under the [CC BY 3.0 license](https://creativecommons.org/licenses/by/3.0/deed.en).
```
